libs/Microsoft.MixedReality.WebRTC.Native/include/interop_api.h

```diff
@@ -209,26 +290,66 @@ enum class TrackKind : uint32_t {
   kDataTrack = 3,
 };
 
-/// Callback fired when a remote track is added to a connection.
-using PeerConnectionTrackAddedCallback = void(MRS_CALL*)(void* user_data,
-                                                         TrackKind track_kind);
-
-/// Callback fired when a remote track is removed from a connection.
-using PeerConnectionTrackRemovedCallback =
-    void(MRS_CALL*)(void* user_data, TrackKind track_kind);
+/// Callback fired when a remote audio track is added to a connection.
```

Comment:
Are these for back compat only? Or a shortcut? I'm surprised to see trackadded callbacks on the peerconnection because if I understand correctly they can only be added via a tranciever.
I expected to see PeerConnection.TranceiverAdded callback and Tranceiver.TrackAdded.

Reply:
No, that's not really how the API works. I know, it's confusing 😉 

Basically the `TrackAdded` and `TrackRemoved` events, which concern only **remote** tracks (receivers), mean that a transceiver started/stopped receiving some media due to an SDP offer/answer being applied. When that happens, a remote track is automatically created/destroyed on the transceiver as needed. Quite often when the transceiver starts receiving, this is a new transceiver that also has just been created by that same offer/answer SDP message. This all happens at the same time during `SetRemoteDescription()`. And the `TrackAdded` event does contain the newly added track **and** the newly added transceiver. `TransceiverAdded` would make less sense, because it would always contain a newly-created receiver track anyway, but that information would not be carried by the name. And if we split the transceiver/track events, then anyway those 2 events would always fire exactly one after the other, which is not very useful.

Note that for local tracks they are not "added" or "removed" on transceivers in the same sense as other objects; instead, a transceiver has a sender "track" slot which can be NULL or any valid local track, and is solely managed by the user. Because of that there is no event associated, since the action is always initiated by the user.

About events naming, this is the official (but outdated) terminology of the Google code, so we can discuss renaming eventually. They were indeed relating to tracks in Plan B, and names remained.

More generally as a mental model things are highly asymmetrical. A naive mental model that I had when reading the standard was that both peers create transceivers, and then the SDP session pairs them 1:1 based on their index (order they have been added). This is wrong and does not work; the actual pairing process is quite a lot more complex, and generally you want the answering peer to wait before adding its transceivers:
1. Peer A adds some transceivers, sets their desired direction (_e.g._ `kSendRecv` or `kSendOnly`).
2. Peer A creates an offer. The offer is based on the existing transceivers and their desired direction at the time of the `CreateOffer()` call.
3. Signaling A => B
4. Peer B applies the offer with `SetRemoteDescription()`. During that time, some transceivers/tracks are created/destroyed based on the offer only. New transceivers are created in `kRecvOnly` mode, always.
5. Peer B optionally adds more transceivers, and/or change the desired direction of the existing ones (_e.g._ from `kRecvOnly` to `kSendRecv`).
6. Peer B creates an answer. Same rules as the offer, except that actual directions advertised are the intersection of the offer and the desired direction, so offer=`kRecvOnly` and desired=`kSendOnly` are incompatible and result in `kInactive`: peer A offered to send, but peer B was only interested in sending too, and refused to receive.
7. Signaling B => A
8. Peer A applies the answer with `SetRemoteDescription()`, same rules as above.

The critical point here is that in step 4. if the answering peer has already some transceivers then there are only 2 possible ways to match those with the `m=` lines from the SDP offer:
- Either both have the same unique name, called `mid`, generated by the implementation;
- Or there is a weird and non-intuitive fallback rule I don't remember and which works only in some cases (in other cases, this rule won't apply, and a new transceiver is created instead).

Because the former requires an external synchronization mechanism (_e.g._ using a data channel, or using the signaling server to exchange some custom messages) to match the `mid` values, and the later is plain confusing and doesn't work in all cases, the best is to start from a clean slate in step 4. (or iterate from a clean previous negotiation). The scheme above is what we do in the Unity integration by the way.